### PR TITLE
Add events to notify of drawer opening / closing

### PR DESCRIPTION
**Pull request by humancopy**
Hey :)

I think it's a good idea to start using events in Timber instead of callbacks to make it easy to tap into interface changes. I also want to add events to the ajax_cart.js file, what you think?

### CODE REVIEW

**Comment by cshold** [comment]
I agree this is a great approach. Are you able to update the docs to include these as well?

**Comment by humancopy** [comment]
Yes, would be happy to do it. I'll add it to this pull request. :)

**Comment by humancopy** [comment]
I added more triggers and fixed the namespace. I've implemented this in a few projects and it's been working great.

Let me know what we can do to push this forward, I'm happy to help with anything needed. 

Have a beautiful weekend.

Peace ;)

**Comment by Cam** [comment]
Awesome. This is exactly what I was talking about a little while back. Thanks heaps @humancopy 😃

File: assets/ajax-cart.js.liquid [hunk 1]
```diff
@@ -72,18 +78,23 @@ ShopifyAPI.onError = function(XMLHttpRequest, textStatus) {
     - Allow custom error callback
 ==============================================================================*/
 ShopifyAPI.addItemFromForm = function(form, callback, errorCallback) {
-  var params = {
+  var $body = $('body'),
```
Comment (cshold) [inline assets/ajax-cart.js.liquid]: A bit of a nitpick, but can you change this to `$(document.body)` - it's slightly faster ([source](https://www.sitepoint.com/jquery-body-on-document-on/)).

File: assets/ajax-cart.js.liquid [hunk 2]
```diff
@@ -39,20 +39,26 @@ ShopifyAPI.onCartUpdate = function(cart) {
...
 ShopifyAPI.updateCartNote = function(note, callback) {
-  var params = {
+  var $body = $('body'),
+  params = {
     type: 'POST',
     url: '/cart/update.js',
     data: 'note=' + attributeToString(note),
     dataType: 'json',
+    beforeSend: function() {
+      $body.trigger('updateCartNote.ajaxCart', note);
```
Comment (cshold) [inline assets/ajax-cart.js.liquid]: What's the need for this event, rather than just on success?

File: assets/ajax-cart.js.liquid (hunk 1 above)
Comment (humancopy) [inline assets/ajax-cart.js.liquid]: Done, changed it in a couple of other places also ;)

File: assets/ajax-cart.js.liquid (hunk 2 above)
Comment (humancopy) [inline assets/ajax-cart.js.liquid]: I changed all to before / after / error / complete events. The before trigger I didn't use but thought to add it as I was already there. :)

File: assets/ajax-cart.js.liquid [hunk 3]
```diff
@@ -325,7 +357,7 @@ var ajaxCart = (function(module, $) {
 
   cartCallback = function(cart) {
     $body.removeClass('drawer--is-loading');
-    $body.trigger('ajaxCart.afterCartLoad', cart);
+    $body.trigger('afterCartLoad.ajaxCart', cart);
```
Comment (cshold) [inline assets/ajax-cart.js.liquid]: Can you revert this name. It's the only breaking change I can see in the PR

Comment (cshold) [inline assets/ajax-cart.js.liquid]: Scratch that. I'll update the version number differently instead. All good

File: layout/theme.liquid [hunk 4]
```diff
@@ -424,7 +424,7 @@
         });
       });
 
-      jQuery('body').on('ajaxCart.afterCartLoad', function(evt, cart) {
+      jQuery(document.body).on('ajaxCart.afterCartLoad', function(evt, cart) {
```
Comment (cshold) [inline layout/theme.liquid]: Please update these two lines to the new name `afterCartLoad.ajaxCart`

Comment (humancopy) [inline layout/theme.liquid]: Oops, done ;)

File: assets/ajax-cart.js.liquid [hunk 5]
```diff
@@ -92,40 +106,55 @@ ShopifyAPI.addItemFromForm = function(form, callback, errorCallback) {
...
 ShopifyAPI.getCart = function(callback) {
+  $(document.body).trigger('beforeGetCart.ajaxCart');
   jQuery.getJSON('/cart.js', function (cart, textStatus) {
     if ((typeof callback) === 'function') {
       callback(cart);
     }
     else {
       ShopifyAPI.onCartUpdate(cart);
     }
+    $(document.body).trigger('afterGotCart.ajaxCart', cart);
```
Comment (cshold) [inline assets/ajax-cart.js.liquid]: Last change, switch this to `afterGetCart`. Realizing the naming issues as I'm writing up some new docs

Comment (humancopy) [inline assets/ajax-cart.js.liquid]: No worries, done... Let me know if need any help with docs. :)

File: assets/ajax-cart.js.liquid [hunk 6]
```diff
@@ -354,6 +386,7 @@ var ajaxCart = (function(module, $) {
 
       // If it has a data-line, update the cart.
       // Otherwise, just update the input's number
+      $body.trigger('adjustCart.ajaxCart', [this, line, qty]);
```
Comment (cshold) [inline assets/ajax-cart.js.liquid]: Can remove this. `updateQuantity` (the function right after this) calls `ShopifyAPI.changeItem`, which already has triggers fired (`before/after/error/complete ChangeItem`)

File: assets/ajax-cart.js.liquid [hunk 7]
```diff
@@ -375,6 +408,7 @@ var ajaxCart = (function(module, $) {
 
       // If it has a data-line, update the cart
       if (line) {
+        $body.trigger('adjustCart.ajaxCart', [this, line, qty]);
```
Comment (cshold) [inline assets/ajax-cart.js.liquid]: Remove this line as well, as per the previous comment.

File: assets/ajax-cart.js.liquid (hunk 6 above)
Comment (humancopy) [inline assets/ajax-cart.js.liquid]: updateQuantity is not always getting triggered which is why I put this one here

File: assets/ajax-cart.js.liquid (hunk 7 above)
Comment (humancopy) [inline assets/ajax-cart.js.liquid]: Another reason is access to the element which isn't available inside updateQuantity

Comment (cshold) [inline assets/ajax-cart.js.liquid]: The entire cart reprints, so having the element available won't do much help. Open to a rebuttal though.

**Comment by cshold** [comment]
Alright, we're almost there.

Let's remove:
- `beforeSubmit.ajaxCart` | duplicate of event in `ShopifyAPI.addItemFromForm`
- `afterSubmit.ajaxCart` | duplicate, same as above
- `itemError.ajaxCart` | duplicate, `ShopifyAPI.addItemFromForm` triggers `errorAddItem.ajaxCart` regardless if an error callback is sent
- `adjustCart.ajaxCart` | duplicate of events triggered in `ShopifyAPI.changeItem`
  - My argument with this one is that if `ShopifyAPI.changeItem` isn't called, the events aren't needed at all. Every other event is tied to an ajax call except this one.

File: assets/ajax-cart.js.liquid (hunk 7 above)
Comment (humancopy) [inline assets/ajax-cart.js.liquid]: I was using it mainly for when the quantity in the cart was changed and then to easily get the id of the variant but that can easily be done using line for the position of the item in the cart and grabbing it from the input ... So I'll remove this, you're right it's just redundant and confusing... :)

**Comment by humancopy** [comment]
Removed redundant events. Docs look really good, well done! :)

**Comment by cshold** [comment]
Thanks for the updates. Did a bunch of testing and it looks in order. The only duplicate events triggered are `afterCartLoad.ajaxCart` and `afterGetCart.ajaxCart` — but I'm happy to leave both. The second one is in the `ShopifyAPI.getCart` function itself, firing no matter what callback is sent, while the first one is specific to Timber's drawer implementation. 

Will give the docs a quick review the merge both. Cheers to putting this together.

**Comment by cshold** [comment]
Last ask, could you squash the commits down to one. [Here are the steps](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/).

**Comment by humancopy** [comment]
Unfortunately I can't squash because I merged stuff from the Shopify/Timber as I needed to keep it up to date cause I was using it for some projects, sorry. If you have an idea how to resolve this I'm happy to follow... ;)

**Comment by cshold** [comment]
No worries, more of a nice to have than anything.

**Pull request merged**